sympy/assumptions/assume.py

```diff
@@ -211,6 +307,131 @@ def eval(self, expr, assumptions=True):
         return res
 
 
+class AskHandlerClass(Dispatcher):
+    """
+    Class for handler which evaluates the ``AppliedPredicate``.
+
+    Explanation
+    ===========
+
+    This class dispatches various types, and return fuzzy boolean values
+    when called.
+    When the dispatched function return ``None``, the next function for
+    the signature is queried until some other value is returned, or no
+    more registered function is left.
+
+    Parameters
+    ==========
+
+    name : str
+
+    doc : str, optional
+
+    base : tuple of AskHandlerClass, optional
+        Base dispatchers, which inherites the dispatched functions.
+
+    Examples
+    ========
+
+    Generate a base dispatcher ``MyHandler``.
+
+    >>> from sympy import Symbol, Add, AskHandlerClass
+    >>> from sympy.abc import x
+    >>> MyHandler = AskHandlerClass('myhandler')
+    >>> @MyHandler.register(Symbol)
+    ... def _(expr, assumptions):
+    ...     return True
+    >>> @MyHandler.register(Add)
+    ... def _(expr, assumptions):
+    ...     return False
+    >>> MyHandler(x, assumptions=True)
+    True
+    >>> MyHandler(x+1, assumptions=True)
+    False
+    >>> print(MyHandler(2*x, assumptions=True)) # unregistred type
+    None
+
+    Generate new dispatcher ``MyHandler2``, which uses ``MyHandler`` as
+    base dispatcher
+
+    >>> MyHandler2 = AskHandlerClass('myhandler2', base=MyHandler)
+    >>> @MyHandler2.register(Add)
+    ... def _(expr, assumptions):
+    ...     return True
+    >>> MyHandler2(x, assumptions=True) # refer to base
+    True
+    >>> MyHandler2(x+1, assumptions=True)
+    True
+
+    """
+
+    def __init__(self, name, doc=None, base=()):
+        super().__init__(name, doc)
+        if not type(base) is tuple:
+            base = (base,)
+        for b in reversed(base):
+            self.funcs.update(b.funcs)
+        self.ordering = ordering(self.funcs)
+        self.base = base
+
+    def __call__(self, *args, **kwargs):
+        # If `None` is returned, search for next function
+        # If all signature return `None`, return `None`
+        types = tuple([type(arg) for arg in args])
+
+        try:
+            func = self._cache[types]
+        except KeyError:
+            func = self.dispatch(*types)
+            self._cache[types] = func
```

Comment:
It looks to me like this is reimplementing stuff that multipledispatch already does. Why can't we just use the normal multiple dispatch API here?

Reply:
Good point. Previous APIs didn't seem flexible enough to support this, so I introduced `get_func()` method to `Dispatcher` in commit aa8914d. Both `Dispatcher.__call__()` and `AskHandlerClass.__call__()` now use that method. Please let me know your thought on this. Thanks!

Reply:
Why is any change needed for multipledispatch here? What's wrong with the normal multipledispatch API?

Ideally we wouldn't make too many changes to multipledispatch because it's a copy of an external module:
https://pypi.org/project/multipledispatch/

At some point I think it would be better to just depend on the other multipledispatch module and remove sympy's copy.

Reply:
That's because handler needs to search for next function if `None` is returned. That's the design of original handler - see comments in `Predicate.deprecated_eval()` method. `Dispatcher` does not do this. It does not search higher classes and just returns `None`.

Reply:
> Ideally we wouldn't make too many changes to multipledispatch because it's a copy of an external module:

Then I'll revert aa8914d.